[video_player] Add video caching to video player

TimHoogstrate opened the pull request:
This resolves feature issue: https://github.com/flutter/flutter/issues/28094

Added caching functionality to video_player, video_player_android, video_player_avfoundation. Changed the platform interface.


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

ditman commented:
The CLA robot is happy, but there's still some CI checks that are broken on this branch and that will need to be taken care of before we can move forward (feel free to ask in this PR if any of the checks have cryptic errors, it happens some times!)

ditman commented:
For the remaining failing check, take a look at this:

* https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changing-federated-plugins (Especially the bit about `make-deps-path-based`).

(Also, please remember to undo the changes in `video_player_web`, if this is only updating metadata, it isn't needed, and it'll make your PR much simpler to manage!)

camsim99 commented on packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/CacheDataSourceFactory.java:
```diff
@@ -0,0 +1,46 @@
+import com.google.android.exoplayer2.upstream.cache.CacheDataSource;
+import com.google.android.exoplayer2.upstream.cache.SimpleCache;
+
+public class CacheDataSourceFactory implements DataSource.Factory {
```
Nit: can you add a javaoc explaining what this class is for?

TimHoogstrate replied:
I've added them

camsim99 commented on packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoCache.java:
```diff
@@ -0,0 +1,59 @@
+  public static void clearVideoCache(@NonNull Context context) {
+    try {
+      File dir = new File(context.getCacheDir(), cacheFolder);
+      deleteDir(dir);
```
If deleting fails, should we let the developer know or throw an exception?

TimHoogstrate replied:
@camsim99 Let me discuss this with the team. I absolutely agree that informing the developer is important. But I'm not sure how to implement this at the moment. Brb.

TimHoogstrate replied:
Oke, I think I'll implement some sort of ClearCacheMessageResponse with a boolean (hasSucceed). In future, we could add some additional information to the ClearCacheMessageResponse. e.g. error messages or additional instructions. Agree?

camsim99 replied:
That makes sense to me because the boolean would allow to error to be handled on the Dart side

TimHoogstrate replied:
Done

camsim99 reviewed (commented):
Reviewed the Android part!

hellohuanlin commented on packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m:
```diff
@@ -176,6 +178,250 @@ - (void)testSeekToInvokesTextureFrameAvailableOnTextureRegistry {
+      OCMProtocolMock(@protocol(FlutterTextureRegistry));
+  NSObject<FlutterPluginRegistry> *registry =
+      (NSObject<FlutterPluginRegistry> *)[[UIApplication sharedApplication] delegate];
+  NSObject<FlutterPluginRegistrar> *registrar = [registry registrarForPlugin:@"CantCachemp4"];
```
and should it be "can cache" or "cant cache"?

TimHoogstrate replied:
modified

hellohuanlin commented on packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m:
```diff
@@ -176,6 +178,250 @@ - (void)testSeekToInvokesTextureFrameAvailableOnTextureRegistry {
+  FLTIsSupportedMessage *response = [videoPlayerPlugin isCacheSupportedForNetworkMedia:message
+                                                                                 error:&error];
+  XCTAssertNil(error);
+  XCTAssertTrue(response.isSupported.boolValue);
```
Looks like it's just checking if `isCacheSupportedForNetworkMedia ` boolean, rather than checking whether it actually performs the caching (e.g. saving data to file, etc). 

How about renaming to something like: `testIsCacheSupported_yesForMP3`

TimHoogstrate replied:
I've renamed the tests

hellohuanlin commented on packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m:
```diff
@@ -176,6 +178,250 @@ - (void)testSeekToInvokesTextureFrameAvailableOnTextureRegistry {
+  XCTAssertFalse(response.isSupported.boolValue);
+}
+
+- (void)testCantCachOGG {
```
I think 1-2 common unsupported formats should be good enough - you can't enumerate all of them anyways. 

May make sense to put them into the same test function.

hellohuanlin commented on packages/video_player/video_player_avfoundation/example/lib/main.dart:
```diff
@@ -119,8 +119,7 @@ class _BumbleBeeRemoteVideoState extends State<_BumbleBeeRemoteVideo> {
     _controller = MiniController.network(
-      'https://flutter.github.io/assets-for-api-docs/assets/videos/bee.mp4',
-    );
+        'https://flutter.github.io/assets-for-api-docs/assets/videos/bee.mp4');
```
revert

TimHoogstrate replied:
Reverted, forgot the comma, therefore it was modified

hellohuanlin commented on packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheAction.h:
```diff
@@ -0,0 +1,16 @@
+
+typedef NS_ENUM(NSInteger, CacheType) { CacheTypeLocal = 0, CacheTypeRemote };
+
+@interface CacheAction : NSObject
```
need prefix for type names (CacheAction and CacheType)

TimHoogstrate replied:
I've added the prefixes.... And I also added the prefixes to the class names.

hellohuanlin commented on packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheAction.h:
```diff
@@ -0,0 +1,16 @@
+
+#import <Foundation/Foundation.h>
+
+typedef NS_ENUM(NSInteger, CacheType) { CacheTypeLocal = 0, CacheTypeRemote };
```
what does local vs remote mean? can you comment here?

TimHoogstrate replied:
Changed the types

hellohuanlin reviewed (changes requested):
For iOS part: 

I haven't gone through the whole PR yet. Please address a few things before I do a full review: 

1. Unit tests on iOS native side (`XCTest`s, not the dart tests) - this is a pretty complicated PR, but it seems the only XCTest I can find is a trivial function ([isCacheSupported](https://github.com/flutter/packages/pull/4526/files#diff-85f606af958fc0025baafe608f631ef8622a8f35e583c1e4f3b5f99cfdc659deR681-R689)). Other than that, the whole logic is almost untested. 

2. Please add doc comments to the header files per our style guide. It will make my review easier. 

3. I reviewed the [previous PR](https://github.com/flutter/packages/pull/4283/files) that has been closed. I noticed that some comments there are not addressed yet.

camsim99 reviewed (commented):
To follow up on @hellohuanlin's review (and address something I didn't say in my review), unit tests will need to be added for the added logic on the Java side.

TimHoogstrate commented on packages/video_player/video_player_android/pigeons/messages.dart:
```diff
@@ -24,6 +24,21 @@ class LoopingMessage {
   bool isLooping;
 }
 
+class IsSupportedMessage {
```
@camsim99, good idea

TimHoogstrate replied:
Modified to IsSupportedMessageResponse

TimHoogstrate commented on packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/CacheDataSourceFactory.java:
```diff
@@ -0,0 +1,46 @@
+        defaultDatasourceFactory.createDataSource(),
+        new FileDataSource(),
+        new CacheDataSink(downloadCache, maxFileSize),
+        CacheDataSource.FLAG_BLOCK_ON_CACHE | CacheDataSource.FLAG_IGNORE_CACHE_ON_ERROR,
```
@camsim99 I've added some information. They are added for performance improvements.

TimHoogstrate commented on packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayerPlugin.java:
```diff
@@ -154,6 +158,11 @@ public void initialize() {
+      boolean isSupported = isCacheSupported(Uri.parse(arg.getUri()));
+      if (isSupported) {
+        player.setMaxCacheSize(arg.getMaxCacheSize());
+        player.setMaxFileSize(arg.getMaxFileSize());
```
good idea, I wanted to follow the same structure as other options. But this sounds like a proper solution

TimHoogstrate commented on packages/video_player/video_player/lib/video_player.dart:
```diff
@@ -545,6 +568,26 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
+    return _applyIsCacheSupported(url);
+  }
+
+  Future<bool> _applyIsCacheSupported(String url) async {
```
I've combined them through the app. So also on other parts

TimHoogstrate commented on packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m:
```diff
@@ -128,6 +128,7 @@ - (void)testBlankVideoBugWithEncryptedVideoStreamAndInvertedAspectRatioBugForSom
                 uri:@"https://flutter.github.io/assets-for-api-docs/assets/videos/bee.mp4"
         packageName:nil
          formatHint:nil
+        enableCache:[NSNumber numberWithInt:0]
```
@hellohuanlin Yes, but... Pigeon generated a NSNumber, so I supposed to use it throughout the objective C code. Is it better to replace it for a bool?

TimHoogstrate replied:
Done

TimHoogstrate commented on packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheAction.h:
```diff
@@ -0,0 +1,16 @@
+- (instancetype)initWithActionType:(CacheType)cacheType range:(NSRange)range;
+
+@property(nonatomic) CacheType cacheType;
+@property(nonatomic) NSRange range;
```
Oke, I'll add doc comments

TimHoogstrate replied:
done

ditman commented:
Unsubbing, please remember to revert the changes to `video_player_web` (or adding the required (noop?) behavior there)

hellohuanlin commented:
unsubscribing too - please request my review again when it's ready.

stuartmorgan commented on packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m:
```diff
@@ -176,6 +212,247 @@ - (void)testSeekToInvokesTextureFrameAvailableOnTextureRegistry {
   XCTAssertEqual([player position], 1234);
 }
 
```
We should have individual unit test files for each specific cache class, in addition to these higher-level native tests.

stuartmorgan commented on packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m:
```diff
@@ -128,6 +128,41 @@ - (void)testBlankVideoBugWithEncryptedVideoStreamAndInvertedAspectRatioBugForSom
+}
+
+- (void)testInitWithCache {
+  // This is to fix 2 bugs: 1. blank video for encrypted video streams on iOS 16
```
I'm not sure what this comment means in this context; a test can't be fixing a bug.

TimHoogstrate replied:
The comment was copied from above. it is now removed.

stuartmorgan commented on packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheAction.h:
```diff
@@ -0,0 +1,30 @@
+
+#import <Foundation/Foundation.h>
+
+typedef enum { CacheTypeLocal = 0, CacheTypeRemote } CacheType;
```
These need declaration comments, per go/objc-style#declaration-comments

stuartmorgan commented on packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheConfiguration.h:
```diff
@@ -0,0 +1,89 @@
+
+/*!
+ Provide functionality related to caching configuration and content download statistics. Let's go
+ through each class and its methods:
```
Many of the comments in this file are incomplete, or don't follow the style guide (e.g., are sentence fragements).

stuartmorgan commented on packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheConfiguration.h:
```diff
@@ -0,0 +1,89 @@
+@interface CacheConfiguration : NSObject <NSCopying>
+
+/*!
+ return configuration based on the provided filePath
```
This comment doesn't actually tell me what this does or how to use it. What is a configuration file path? What's the relationship to the provided path? What even is the provided path supposed to be?

stuartmorgan commented on packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheConfiguration.h:
```diff
@@ -0,0 +1,89 @@
+ Provide functionality related to caching configuration and content download statistics. Let's go
+ through each class and its methods:
+ */
+@interface CacheConfiguration : NSObject <NSCopying>
```
The documentation for these classes would benefit from a higher-level description of how this class fits into an overall cache system. E.g., I can guess from some of the APIs that this is probably a per-cached-item configuration, but that's not at all obvious from the class-level docs.

TimHoogstrate replied:
added additional documentation

stuartmorgan commented on packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheManager.h:
```diff
@@ -0,0 +1,31 @@
++ (CacheConfiguration *)cacheConfigurationForURL:(NSURL *)url error:(NSError **)error;
+
+/*!
+ removes all (downloading)files in cache directory
```
Why would clearing the cache only clear in-progress downloads?

TimHoogstrate replied:
it does not. I'll change the documentation, it seems unclear.

stuartmorgan commented on packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheSessionManager.h:
```diff
@@ -0,0 +1,25 @@
+ named downloadQueue, which can be used to handle download operations efficiently. This type of
+ design is commonly used for managing background tasks, such as downloading and caching content for
+ a video player, to ensure proper concurrency and control over the operations. The singleton pattern
+ guarantees that all parts of the application that need access to the download queue use the same
```
Why do we want to enforce that guarantee via a singleton? If there's a top-level cache manager class, then can't it provide the session manager to individual classes?

Using singletons rather than DI to share when it's not absolutely required that there be only one of something globally (which I wouldn't expect to be the case here, but maybe I'm not understanding the scope of what this does) makes unit testing much harder.

stuartmorgan commented on packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheSessionManager.h:
```diff
@@ -0,0 +1,25 @@
+ @Property downloadQueue
+ NSOperationQueue that is reponsible for download operations.
+ */
+@property(nonatomic, strong, readonly) NSOperationQueue *downloadQueue;
```
In fact, if it weren't a singleton then the whole class could go away, and you could just pass the queue to client classes.

TimHoogstrate replied:
The singleton pattern guarantees that all parts of the application that need access to the download queue use the same instance, avoiding duplication and synchronization issues.

stuartmorgan commented on packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/ContentCacheWorker.h:
```diff
@@ -0,0 +1,73 @@
+ @property setupError
+ error callback, returns error when setup fails
+ */
+@property(nonatomic, strong, readonly) NSError *setupError;
```
Why is this a property rather than an out param? The latter is much more idiomatic ObjC.

TimHoogstrate replied:
I've changed it

stuartmorgan reviewed (changes requested):
I did an initial pass, but I didn't get into all of the Obj-C cache implementation since I was finding it hard to understand how the pieces fit together at a high level before reviewing the implementations. Some high-level doc comments would help, but I think it would also help a whole lot to do a short [design document](https://github.com/flutter/flutter/wiki/Design-Documents) for the iOS cache that explains the design and the pieces at a high level.

TimHoogstrate commented on packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayerPlugin.java:
```diff
@@ -171,6 +180,10 @@ public void setLooping(@NonNull LoopingMessage arg) {
     player.setLooping(arg.getIsLooping());
   }
 
+  public void clearCache(@NonNull Messages.ClearCacheMessage msg) {
```
Done

TimHoogstrate commented on packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayerPlugin.java:
```diff
@@ -248,4 +268,19 @@ void stopListening(BinaryMessenger messenger) {
+  @Nullable
+  public static String getMimeType(@NonNull String url) {
+    String type = null;
+    String extension = MimeTypeMap.getFileExtensionFromUrl(url);
```
Added additional check

TimHoogstrate commented on packages/video_player/video_player/lib/video_player.dart:
```diff
@@ -348,6 +354,21 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
 
+  /// **iOS only**. Will set is caching is true (enabled) or false (disabled), default (disabled).
+  /// currently only mimetypes: video/mp4 and audio/flac files can be cached. For other mimetypes this setting is ignored.
+  /// detection with whatever is set here.
```
Changed to capitals

TimHoogstrate commented on packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart:
```diff
@@ -127,14 +138,16 @@ class DataSource {
+      this.asset,
+      this.package,
+      this.httpHeaders = const <String, String>{},
+      this.maxCacheSize,
```
I agree, lets see if I can modify this to something like cache settings.

TimHoogstrate commented on packages/video_player/video_player/lib/video_player.dart:
```diff
@@ -275,6 +278,9 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
     this.formatHint,
+    this.maxCacheSize,
+    this.maxFileSize,
+    this.enableCache,
```
I've added it into the existing VideoPlayerOptions.

TimHoogstrate replied:
I tried adding it to the existing VideoPlayerOptions. Let me know what you think of it.

TimHoogstrate commented on packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart:
```diff
@@ -78,6 +78,17 @@ abstract class VideoPlayerPlatform extends PlatformInterface {
+  }
+
+  /// Clears the cached videos.
+  Future<bool> clearCache(int textureId) {
```
that is actually correct, I think this is a leftover from the past. :)

TimHoogstrate commented on packages/video_player/video_player/lib/video_player.dart:
```diff
@@ -264,8 +264,11 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
+  /// null.
+  /// **Android only**: The [maxCacheSize] option allows the caller to override
+  /// **Android only**: The [maxFileSize] option allows the caller to override
+  /// **iOS only**: The [enableCache] option allows the caller to override
```
I use the videoPlayerOptions object and moved the docs in there.

TimHoogstrate commented on packages/video_player/video_player/lib/video_player.dart:
```diff
@@ -348,6 +357,21 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
   final VideoFormat? formatHint;
 
+  /// **iOS only**. Will set is caching is true (enabled) or false (disabled), default (disabled).
+  /// Currently only mimetypes: video/mp4 and audio/flac files can be cached. For other mimetypes this setting is ignored.
```
I've added it to the VideoPlayerOptions

TimHoogstrate commented on packages/video_player/video_player/lib/video_player.dart:
```diff
@@ -545,6 +571,22 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
   }
 
+  /// Returns if caching is supported for network media.
+  Future<bool> isCacheSupportedForNetworkMedia(String url) async {
```
Changed throughout the app

TimHoogstrate commented on packages/video_player/video_player_web/lib/video_player_web.dart:
```diff
@@ -127,6 +127,19 @@ class VideoPlayerPlugin extends VideoPlayerPlatform {
+  /// Returns if cache is supported.
+  @override
+  Future<bool> isCacheSupportedForNetworkMedia(String url) async {
+    // caching is by default supported by most browser. It is however a function of the browser.
```
I've added a caching description to the README of web.

TimHoogstrate commented on packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheAction.h:
```diff
@@ -0,0 +1,30 @@
+
+- (instancetype)initWithActionType:(CacheType)cacheType range:(NSRange)range;
+
+/*!
```
Changed it

TimHoogstrate commented on packages/video_player/video_player_avfoundation/CHANGELOG.md:
```diff
@@ -9,6 +13,7 @@
 
 ## 2.4.7
 
+* Adds caching functionality for network video.
```
removed the line

camsim99 reviewed (commented):
Android part looks good to me; just needs unit tests for the added Java functionality/classes!

stuartmorgan commented on packages/video_player/video_player_avfoundation/ios/Classes/Cache/ResourceLoader/FVPResourceLoader.h:
```diff
@@ -0,0 +1,34 @@
+@property(nonatomic, strong, readonly) NSURL *url;
+@property(nonatomic, weak) id<FVPResourceLoaderDelegate> delegate;
+
+- (instancetype)initWithURL:(NSURL *)url error:(NSError *)error;
```
(These comments about init methods apply throughout the ObjC code in the PR.)

stuartmorgan commented on packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/FVPContentCacheWorker.h:
```diff
@@ -0,0 +1,70 @@
+ */
+@interface FVPContentCacheWorker : NSObject
+
+- (instancetype)initWithURL:(NSURL *)url error:(NSError *)error;
```
As with the other class, I'm confused about the meaning of passing an error into this method.

(Relatedly, init methods with params generally need documentation explaining the params.)

TimHoogstrate replied:
If initialisation fails the error contains details about the failure. I've added the documentation

stuartmorgan replied:
Is it intended to be an out parameter? An `NSError` out param needs to be `NSError**`, not `NSError*`, or it won't work. If that's what this is supposed to before, then it seems like there must not be tests of it since it can't work this way.

Also, an `init` method having an error out param is not a pattern I've ever seen before. If there's something that's expected to fail in normal use, then it should be a separate method, not `init*`.

TimHoogstrate replied:
Oke I'll make some adjustments. I don't know why I added it later but it was not like this before.. probably because I was testing something, or working on the unit tests... I'll update the PR

stuartmorgan reviewed (changes requested):
I made it through all but the Obj-C code. I've started on the Obj-C code, but still have a lot to go. Since a number of my comments there so far are going to apply to many of the files though, I'm publishing now (since I won't be able to finish the review today) to get some feedback out now. (I'll try to finish the review tomorrow, but the Obj-C is a lot of code still.)

TimHoogstrate commented on packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart:
```diff
@@ -49,7 +49,8 @@ abstract class VideoPlayerPlatform extends PlatformInterface {
   /// Creates an instance of a video player and returns its textureId.
-  Future<int?> create(DataSource dataSource) {
+  Future<int?> create(
+      DataSource dataSource, VideoPlayerOptions? videoPlayerOptions) {
```
good idea, do you have an idea about a suitable name for the new create method? It would then be something like: 'Future<int?> newCreate(
      DataSource dataSource, CreationParameters? creationParameters) {'

TimHoogstrate replied:
I've made a createWithParameters method

TimHoogstrate commented on packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart:
```diff
@@ -394,6 +409,27 @@ class VideoPlayerOptions {
+  /// Detection with whatever is set here.
+  final bool enableCache;
+
+  /// Will set the size of the total cache. Default maxCacheSize is 0 (no cache enabled).
```
I add a thumbs up to keep track on the "implemented changes"

TimHoogstrate commented on packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart:
```diff
@@ -394,6 +409,27 @@ class VideoPlayerOptions {
+  /// Detection with whatever is set here.
+  final bool enableCache;
+
+  /// Will set the size of the total cache. Default maxCacheSize is 0 (no cache enabled).
```
agree

TimHoogstrate commented on packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoCache.java:
```diff
@@ -0,0 +1,60 @@
+      File dir = new File(context.getCacheDir(), cacheFolder);
+      return deleteDir(dir);
+    } catch (Exception e) {
+      e.printStackTrace();
```
good question, I guess it is because it is not really helpful for the developer but, I could add an error message to the ResponseMessage

TimHoogstrate commented on packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayerPlugin.java:
```diff
@@ -145,6 +152,11 @@ public void initialize() {
     } else {
       Map<String, String> httpHeaders = arg.getHttpHeaders();
+      boolean isSupported = isCacheSupported(Uri.parse(arg.getUri()));
+      if (isSupported) {
```
No, if these "Options" are set then we assume it is enabled.

TimHoogstrate commented on packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoCache.java:
I've added them already before.

kishangohel commented:
@TimHoogstrate @stuartmorgan @tarrinneal 
When can we expect this feature as this really needed feature and most developers switched to official player

TimHoogstrate commented on packages/video_player/video_player_android/pigeons/messages.dart:
```diff
@@ -24,6 +24,21 @@ class LoopingMessage {
+class ClearCacheMessageResponse {
+  ClearCacheMessageResponse(this.hasSucceeded);
+  bool hasSucceeded;
+}
```
Oke I'll update the Pigeon version to the latest and implement the changes

TimHoogstrate replied:
@stuartmorgan, I've updated Pigeon, but I don't exacly understand what you mean. I understand that Pigeon supports primitive returns, however, by adding these primitives to a "class" it makes it more flexible when we want to add additional properties to the class in the future without changing too much. Let me know if I'm correct here or if you would like the modifications.

stuartmorgan replied:
> by adding these primitives to a "class" it makes it more flexible when we want to add additional properties to the class in the future without changing too much

This is a pattern we follow for platform interface methods because those are public API intended for subclassing, so adding parameters there is breaking. The Pigeon interface is entirely internal, so if we wanted additional parameters we would just add them, and in the unlikely event we needed additional return data in a response we would add a method at that point. There's no reason for Pigeon methods to pre-emptively use classes for everything any more than any other internal API.

TimHoogstrate commented:
Dear @stuartmorgan,

I had some issues while merging macOS implementations into my branch. I fixed them and then, while testing, I noticed an issue, that a black video frame was shown in the player en I was running the example app op a real iPhone after clearing the cache. The simulator did not have that issue. I found out that I accidentally removed some code while documenting. Anyway, I fixed it and I had to make some changes for the macOS implementation. For now I omitted the caching feature for macOS. Do you want to have it implemented now or later? I am working on make it buildable again. It builds and tests locally for both MacOS and iOS. So I kicked off a new build and I'll see how I get this buildable on the build server later.

Dear @hellohuanlin,

I added the android tests and when everything is building you can re review the changes. 

Kind regards,

stuartmorgan commented:
> For now I omitted the caching feature for macOS. Do you want to have it implemented now or later?

What's different between macOS and iOS for the cache? I would have thought it would be entirely shared code.

cbenhagen commented:
> What's different between macOS and iOS for the cache? I would have thought it would be entirely shared code.

@TimHoogstrate did you see this question? I would love to see macOS support from the beginning.

stuartmorgan commented on packages/video_player/video_player/lib/video_player.dart:
```diff
@@ -545,6 +547,22 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
+  }
+
+  /// Clears the cache of the player, returns true if succeeded.
+  Future<bool> clearCache() async {
```
@TimHoogstrate Ping on this question, since it's fundamental to the high-level design.

bgenidy replied:
I agree, I think having a cache configuration per network call can be a bit confusing and can lead to unexpected behaviors if the caller changes the option set for different code path in there code.

One approach is to let the caller pass in the cache store per video instead of the config, such as the below example. Passing the cache store would indicate that this video should be cached, if null then no cache.

```dart
VideoPlayerController controller = VideoPlayerController.networkUrl(
        uri,
        videoPlayerOptions: VideoPlayerOptions(
            cacheStore: networkVideoCacheStore),
      );
```

Then to make it easy on consumers a DefaultCacheStore can be provided where users can pass that in; however, if they want to further customize the caching/partitioning of videos being streamed they have the full ability to implement the interface in a different way as long as they abide by the interface contract/exception pattern.

@stuartmorgan/ @TimHoogstrate what are your thoughts on that approach?

bgenidy replied:
One other thought I had on this, is the cached network store can be implemented as a parallel class where the VideoPlayerController exposes a streamed operation and the wrapper class just consumes that stream and caches it while also passing it to the necessary render widget without having to do a double read of the stream info.

TimHoogstrate commented:
@stuartmorgan, I was a bit lost regarding the cache and macOS implementation. The macOS Implementation was not working because it appears that the MobileCoreServices, (that I use to determine the mime and content-type) were not building (available) for MacOS. I migrated the MobileCoreServices to CoreServices (appearing that this was working for macOS). HOWEVER, I was testing the macOS cache feature (thinking that it would work out of the box), and I ran into some runtime errors. Therefore, in my first reaction was to omit the macOS feature for now (to get the tests and builds running). And then we could add those later as a new feature of version. HOWEVER, the tests were failing because I omitted macOS caching (isCacheAvailable  if Mac then return no else yes). It seemed that I was not able to determine the platform or OS as expected... Then I noticed that in Xcode I run a build on "macOS (designed for iPAD)" I dug into this and then i found out that running a build or test from command line like e.g.: `dart run script/tool/bin/flutter_plugin_tools.dart native-test --macos --packages video_player` has a different result than running the same tests in Xcode. I don't fully understand why. But now that the builds are building again and the tests are success I can dive into the macOS feature, and why it gave me runtime errors.

stuartmorgan commented:
> Then I noticed that in Xcode I run a build on "macOS (designed for iPAD)"

That's Catalyst; it's not something we support in Flutter.

TimHoogstrate commented:
@stuartmorgan, macOS is not accepting an url and throws an url not supported exception on macOS. Same is working on iOS. So I am not sure how to proceed. It seems that macOS needs a bit additional work. iOS is working fine. 

Kind regards,

stuartmorgan commented:
> macOS is not accepting an url and throws an url not supported exception on macOS

Without more details and context I'm not sure what that would be. It's fine to disable caching on macOS with a TODO for now though, and we can investigate once mobile is working.

kishangohel commented:
@TimHoogstrate 
On iOS player throws an internet error although the entire video is cached when the internet is turned off or the internet is not available

hellohuanlin commented:
I am checking if this is ready for a second review on iOS part? Feel free to ping me whenever it's ready.

TimHoogstrate commented:
I was able reproduce the internet error. The thing is that I don't know how to resolve it fast because I don't know what throws the error on iOS. It has been working fine but looks like the URLAsset is throwing an error on iOS (in case there is no internet) or on macOS (in case I use caching, because I use a caching path as an url). I'll try to find a solution in the next week. Then I'll fix the merge conflicts too and check if everything is working as expected.

bgenidy reviewed (commented):
@TimHoogstrate @ditman is this PR still planned to merge in? I have a project that Im trying to enable some sort of preemptive caching of videos without caching the full video or doing a double read of the video stream from network to avoid network impact for users.

Happy to be one of the first to test this out if it needs testers, or if the team was going into a different direction would love to learn more.

delfme commented:
It seems this might resolve to a dead end. If that is the case, would it be doable for flutter team to take over or implement from scratch? `better_player` has a cache feature for both ios and android under MIT. It is a few hundred lines of code for each platform. https://github.com/jhomlala/betterplayer/blob/master/ios/Classes/CacheManager.swift

@stuartmorgan we noticed that the hard work on flutter core and impeller migration has been taking the most of the time and core plugins like `camera` and `videoplayer` are a bit left behind. Even issues have been triaged, yet not worked. Do you have any plan to get a few devs helping on them on a daily basis? Thx

stuartmorgan commented:
> would it be doable for flutter team to take over or implement from scratch?

The plan is already for the Flutter team to continue this PR once we have time.

As for the rest of the comment, an individual PR is not the place for meta-discussion about overall team resourcing. A general discussion forum such as the Discord server would be the place for that.

vashworth commented:
Converting to draft since it's not actively being worked on and not ready for review.

stuartmorgan commented:
Update from triage: This is still on our radar, but other work has taken priority so far.

stuartmorgan commented:
Closing for now so this doesn't show up in triage; we don't expect to have time to work on it in the next several months, and interop work using `video_player` would require reworking this further.
